crates/lsp/src/lib.rs

```diff
@@ -383,26 +399,66 @@ impl<L: LSPLang> Backend<L> {
           range,
           new_text: String::from_utf8(edit.inserted_text).unwrap(),
         };
-        let mut changes = HashMap::new();
-        changes.insert(text_doc.uri.clone(), vec![edit]);
-        let edit = Some(WorkspaceEdit {
-          changes: Some(changes),
-          document_changes: None,
-          change_annotations: None,
-        });
-        let action = CodeAction {
-          title: config.message.clone(),
-          command: None,
-          diagnostics: None,
-          edit,
-          disabled: None,
-          kind: Some(CodeActionKind::QUICKFIX),
-          is_preferred: Some(true),
-          data: None,
-        };
-        response.push(CodeActionOrCommand::from(action));
+        changes
+          .entry(text_document.uri.clone())
+          .or_default()
+          .push(edit);
+      }
+    }
+    Some(changes)
+  }
+
+  async fn on_code_action(&self, params: CodeActionParams) -> Option<CodeActionResponse> {
+    let text_doc = params.text_document;
+    let path = text_doc.uri.to_file_path().ok()?;
+    let diagnostics = params.context.diagnostics;
+    let error_id_to_ranges = Self::build_error_id_to_ranges(diagnostics);
+    let mut response = CodeActionResponse::new();
+
+    let code_action = params.context.only.as_ref()?.first();
+
+    // we only handle these code_actions
+    // 1. QuickFix
+    // 2. "source.fixAll" and "source.fixAll.ast-grep"
+    match code_action {
```

Comment:
NIT: we can use `if matches!(xxx)` to simplify this code

Reply:
I questioned chatGPT, and he told me it should be simplified as below. 😂how do you think?
```rust
	let code_actions = params.context.only.as_deref();
    if !matches!(
      code_actions,
      Some(&[
        CodeActionKind::QUICKFIX
          | CodeActionKind::SOURCE_FIX_ALL
          | CodeActionKind::SOURCE_FIX_ALL_AST_GREP,
        ..
      ])
    ) {
      return Some(response);
    }
```

Reply:
I modified it
```rust
  if code_action != CodeActionKind::QUICKFIX
      && code_action != CodeActionKind::SOURCE_FIX_ALL
      && code_action != SOURCE_FIX_ALL_AST_GREP
    {
      return Some(response);
    }
```